src/lib.rs

```diff
@@ -1,29 +1,37 @@
+pub mod args;
+pub mod commands;
+pub mod error;
+
 use std::{ffi::OsString, path::Path};
 
-use commands::{nsec3hash::Nsec3Hash, LdnsCommand};
+use clap::Parser;
 
-pub use self::args::Args;
+use args::Args;
+use commands::{nsec3hash::Nsec3Hash, LdnsCommand};
+use error::Error;
 
-pub mod args;
-pub mod commands;
-pub mod error;
+pub fn parse_args<I: IntoIterator<Item = OsString>, T: Fn() -> I>(
+    args_provider: T,
+) -> Result<Args, Error> {
+    try_ldns_compatibility(args_provider()).or_else(|_| {
+        Args::try_parse_from(args_provider()).map_err(|err| Error::new(err.to_string().as_str()))
+    })
+}
 
-pub fn try_ldns_compatibility<I: IntoIterator<Item = OsString>>(args: I) -> Option<Args> {
+pub fn try_ldns_compatibility<I: IntoIterator<Item = OsString>>(args: I) -> Result<Args, Error> {
     let mut args_iter = args.into_iter();
-    let binary_path = args_iter.next()?;
+    let binary_path = args_iter
+        .next()
+        .ok_or::<Error>("Missing binary name".into())?;
```

Comment:
I think the type annotation is only necessary because the `?` adds an `into` as well, so removing the `.into()` might clean this up a bit.

Reply:
The latest code changes this behaviour, but I noted there still seems to be unwanted error output in some cases, e.g. on `--help`, so I would rather just merge your env PR to main and merge that into this and see what we end up with.